shim/batch.go

```diff
@@ -0,0 +1,69 @@
+// Copyright the Hyperledger Fabric contributors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package shim
+
+import (
+	"github.com/hyperledger/fabric-protos-go-apiv2/peer"
+)
+
+type writeBatch struct {
+	writes map[string]*peer.WriteRecord
+}
+
+func newWriteBatch() *writeBatch {
+	return &writeBatch{
+		writes: make(map[string]*peer.WriteRecord),
+	}
+}
+
+func (b *writeBatch) Writes() []*peer.WriteRecord {
+	if b == nil {
+		return nil
+	}
+
+	var results []*peer.WriteRecord
+	for _, value := range b.writes {
+		results = append(results, value)
+	}
+
+	return results
+}
+
+func (b *writeBatch) PutState(collection string, key string, value []byte) {
+	b.writes[batchLedgerKey(collection, key)] = &peer.WriteRecord{
+		Key:        key,
+		Value:      value,
+		Collection: collection,
+		Type:       peer.WriteRecord_PUT_STATE,
+	}
+}
+
+func (b *writeBatch) PutStateMetadataEntry(collection string, key string, metakey string, metadata []byte) {
+	b.writes[batchLedgerKey(collection, key)] = &peer.WriteRecord{
+		Key:        key,
+		Collection: collection,
+		Metadata:   &peer.StateMetadata{Metakey: metakey, Value: metadata},
+		Type:       peer.WriteRecord_PUT_STATE_METADATA,
+	}
+}
+
+func (b *writeBatch) DelState(collection string, key string) {
+	b.writes[batchLedgerKey(collection, key)] = &peer.WriteRecord{
+		Key:        key,
+		Collection: collection,
+		Type:       peer.WriteRecord_DEL_STATE,
+	}
+}
+
+func (b *writeBatch) PurgeState(collection string, key string) {
+	b.writes[batchLedgerKey(collection, key)] = &peer.WriteRecord{
+		Key:        key,
+		Collection: collection,
+		Type:       peer.WriteRecord_PURGE_PRIVATE_DATA,
+	}
+}
+
+func batchLedgerKey(collection string, key string) string {
+	return prefixStateDataWriteBatch + collection + key
```

Comment:
You have lost the use of `prefixMetaDataWriteBatch`

Reply:
The names of keys for state and metadata can coincide (theoretically). To store them in the database, they are prepended with their own prefixes, but this is on the peer side. To store them in the batches in one map, I created prefixes for them in the form of constants. Perhaps they should be moved here.

Reply:
Good catch. It seems to me that concatenating collection and key names is not sufficient to distinguish different values. What if we have collections `c` and `col`, and values `old` and `d`. Then `c` + `old` is the same as `col` + `d`.

Reply:
```
prefixStateDataWriteBatch + collection + string(rune(minUnicodeRuneValue)) + key
```